src/main/java/net/neoforged/neoforge/registries/ClientRegistryManager.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.registries;
+
+import com.google.common.collect.Sets;
+import com.mojang.logging.LogUtils;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import net.minecraft.ChatFormatting;
+import net.minecraft.client.Minecraft;
+import net.minecraft.core.Registry;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.neoforged.neoforge.network.handling.ConfigurationPayloadContext;
+import net.neoforged.neoforge.network.handling.PlayPayloadContext;
+import net.neoforged.neoforge.network.payload.KnownRegistryDataMapsPayload;
+import net.neoforged.neoforge.network.payload.KnownRegistryDataMapsReplyPayload;
+import net.neoforged.neoforge.network.payload.RegistryDataMapSyncPayload;
+import org.jetbrains.annotations.ApiStatus;
+import org.slf4j.Logger;
+
+@ApiStatus.Internal
+public class ClientRegistryManager {
+    private static final Logger LOGGER = LogUtils.getLogger();
+
+    public static <R> void handleDataMapSync(final RegistryDataMapSyncPayload<R> payload, final PlayPayloadContext context) {
+        context.workHandler().submitAsync(() -> {
+            final BaseMappedRegistry<R> registry = (BaseMappedRegistry<R>) Minecraft.getInstance().level.registryAccess()
+                    .registryOrThrow(payload.registryKey());
+            registry.dataMaps.clear();
+            payload.dataMaps().forEach((attachKey, maps) -> registry.dataMaps.put(RegistryManager.getDataMap(payload.registryKey(), attachKey), Collections.unmodifiableMap(maps)));
+        }).exceptionally(ex -> {
+            context.packetHandler().disconnect(Component.translatable("neoforge.network.data_maps.failed", payload.registryKey().location(), ex.getMessage()));
+            LOGGER.error("Failed to handle registry data map sync: ", ex);
+            return null;
+        });
+    }
+
+    public static void handleKnownDataMaps(final KnownRegistryDataMapsPayload payload, final ConfigurationPayloadContext context) {
+        record MandatoryEntry(ResourceKey<Registry<?>> registry, ResourceLocation id) {}
+        final Set<MandatoryEntry> ourMandatory = new HashSet<>();
+        RegistryManager.getDataMaps().forEach((reg, values) -> values.values().forEach(attach -> {
+            if (attach.mandatorySync()) {
+                ourMandatory.add(new MandatoryEntry(reg, attach.id()));
+            }
+        }));
```

Comment:
Can we put this in a private static helper method to basically go from data maps -> set<Mandatory>? So we don't duplicate it between calculating our and their mandatory.

Reply:
The map types between `getDataMaps()` and the payload are different, it doesn't make sense as it would mostly be functions converting from one thing to another.